### DISCUSSION TITLE
How does fwupd update the FMP UEFI capsule file ( cab file) generated from edk2?

### DISCUSSION THREAD

**Discussion by gordontcp** [fwupd/fwupd#5248] [Q&A]
I used the example of edk2 platform to generate Red.16.cap and Red.16.cab, the source code is as follows:
`https://github.com/tianocore/edk2-platforms/tree/master/Platform/Intel/Vlv2TbltDevicePkg`
Where Red.16.cap is a dummy test example and does not actually update the firmware.
I execute the following command in UEFI shell to update fmp UEFI capsule successfully:
 ` capsuleapp Red.16.cap`
My system is as follows:
```
Minnowboard Max
UEFI x64
Linux fedora 6.0.5-200.fc36.x86_64
```
I execute following command n fedora, 
`sudo fwupdmgr install Red.16.cab `
After the end, it shows the following messages:
`Successfully installed firmware.Restart Now?[y|N] `
After choosing to restart, I see the following error message under UEFI:
```
Unknown Capsule Guid - 72E2945A-00DA-448E-9AA7-075AD840F9D4
WARNING: QueryCapsuleCapabilities failed, assuming EfiResetWarm: Unsupported
Unknown Capsule Guid - 72E2945A-00DA-448E-9AA7-075AD840F9D4
WARNING: Could not apply capsule update: Unsupported
WARNING: Could not apply capsules: Unsupported
```
It can be seen that the fwupd update fw did not succeed。

I execute `‘sudo fwupdmgr get-devices’ `on fedora and got the following messages:
```
─UEFI Device Firmware:
│ │ Device ID: f95c9218acd12697af946874bfe4239587209232
│ │ Previous version: 1
│ │ Update State: Needs reboot
│ │ Last modified: 2022-07-14 00:19
│ │ GUID: 72e2945a-00da-448e-9aa7-075ad840f9d4
│ │ Device Flags: • Internal device
│ │ • Updatable
│ │ • System requires external power source
│ │ • Needs a reboot after installation
│ │ • Device is usable for the duration of the update
│ │
│ └─ New version: 16
│ License: Unknown
│ Description:
│ The vendor did not supply any release notes.
```
As can be seen from GenCapsuleAll.py of edk2:
`GenCapsuleSampleDevice('Red','72e2945a-00da-448e-9aa7-075ad840f9d4',0x00000010,0x00000000, CapsulesPath, CapsulesSubDir)`
The Capsule Guid can be found by fwupd, but the system reboot shows Unknown Capsule Guid’.

By the way, after updating Red.16.cab, I have successfully restarted Linux, but after restarting, it shows that it needs to be restarted.
My questions are as follows:
1. Update Red.16.cap in UEFI shell, restarts system and it got the following messages
`Setting BootMode to BOOT_ON_S3_RESUME`
And restart after fwupd update Red.16.cab, see the following UEFI message:
`Setting BootMode to BOOT_WITH_FULL_CONFIGURATION`
It can be seen that the mode of the two is different after UEFI restarts. Is it correct that fwupd does not enter S3_RESUME mode after restarting?
2. Can the cab file generated in edk2 be updated directly in fwupd? Or what to modify before executing fwupd?
3. Does fwupd have other samples to test UEFI capsule(cab) update? Are there detailed operating procedures?
Any suggestion is highly appreciated!
Thanks!

**Comment by hughsie** [comment]
On Tue, 8 Nov 2022 at 11:42, gordontcp ***@***.***> wrote:
> WARNING: QueryCapsuleCapabilities failed, assuming EfiResetWarm: Unsupported
> WARNING: Could not apply capsule update: Unsupported
> WARNING: Could not apply capsules: Unsupported

This is telling us that the runtime interface doesn't implement the
required methods. e.g. we call RT->QueryCapsuleCapabilities and get
the EFI_UNSUPPORTED return code back. What's slightly confusing is
that applying the new capsule uses the *old* firmware to apply it, and
so even if the new capsule adds capsule update support we can't apply
it until the existing-firmware-you're-upgrading-from also supports
capsule updates.

Does that help at all?

Richard

**Reply by gordontcp** [reply]
Hello.
Thank you for your reply.
As mentioned, RT->QueryCapsuleCapabilities gets EFI_UNSUPPORTED.  Currently my cap file can be updated normally in UEFI shell, do I still need to modify code to respond `RT->QueryCapsuleCapabilities` separately when using fwupd?
Should I make changes in` \fwupd\plugins\uefi-capsule` to respond?
Thanks!

**Reply by hughsie** [reply]
>  Currently my cap file can be updated normally in UEFI shell

How are you updating in the UEFI shell? Using something like AUEFI as a loader?

**Reply by gordontcp** [reply]
As above, I developed FMP DXE driver from following code:
`https://github.com/tianocore/edk2-platforms/tree/master/Platform/Intel/Vlv2TbltDevicePkg`

I execute the following command in UEFI shell to update fmp UEFI capsule successfully:
`capsuleapp Red.16.cap`
where Red.16.cap is a dummy test example.
My system is as follows:

```
Minnowboard Max
UEFI x64
Linux fedora 6.0.5-200.fc36.x86_64
```
Thanks!

**Reply by hughsie** [reply]
I don't know what `capsuleapp` is I'm afraid. @superm1 any ideas?

**Reply by superm1** [reply]
It looks like it's the name of the binary EFI app that is generated?
https://github.com/tianocore/edk2-platforms/blob/aa3f6fd542e99dde4206537b095f1a2201275e75/Platform/Intel/Vlv2TbltDevicePkg/Feature/Capsule/GenerateCapsule/GenCapsuleAll.py#L368

**Reply by superm1** [reply]
I think it's this: https://github.com/tianocore/edk2/blob/master/MdeModulePkg/Application/CapsuleApp/CapsuleApp.c

**Reply by superm1** [reply]
The thing that stands out to me is that application calls the same RT service....
https://github.com/tianocore/edk2/blob/master/MdeModulePkg/Application/CapsuleApp/CapsuleApp.c#L969
https://github.com/fwupd/fwupd-efi/blob/main/efi/fwupdate.c#L457

The arguments are the same, so there must be some nuance that is different.

**Reply by gordontcp** [reply]
Thanks for reply.
1.>>I think it's this: https://github.com/tianocore/edk2/blob/master/MdeModulePkg/Application/CapsuleApp/CapsuleApp.c
Answer:Yes!

2.>>The arguments are the same, so there must be some nuance that is different.
Answer:
As far as I know, the following items are not the same：
Update Red.16.cap in UEFI shell, restarts system and it got the following messages
`Setting BootMode to BOOT_ON_S3_RESUME`
And restart after fwupd update Red.16.cab, see the following UEFI message:
`Setting BootMode to BOOT_WITH_FULL_CONFIGURATION`
It can be seen that the mode of the two is different after UEFI restarts. Is it correct that fwupd does not enter S3_RESUME mode after restarting?

**Reply by superm1** [reply]
It looks like there are two different flows in there actually.  It will query whether or not to call reset by what's in the capsule header.  I think we should be able to port that flow into fwupd-efi as the code I linked is BSD.

**Comment by gordontcp** [comment]
`>> I think we should be able to port that flow into fwupd-efi as the code I linked is BSD.`
Hi, may I ask, is there any plan to add the above items to the new version of fwupd?

**Reply by hughsie** [reply]
>  Hi, may I ask, is there any plan to add the above items to the new version of fwupd?

Not a plan from me, but do you think you could patch https://github.com/fwupd/fwupd-efi/blob/main/efi/fwupdate.c for this support?
